overlord/fdestate/fdestate.go

```diff
@@ -403,3 +404,71 @@ func MockSecbootGetPCRHandle(f func(devicePath, keySlot, keyFile string, hintExp
 		secbootGetPCRHandle = old
 	}
 }
+
+// KeyslotTarget uniquely identifies a target key slot by
+// its container role and name.
+type KeyslotTarget struct {
+	ContainerRole string `json:"container-role"`
+	Name          string `json:"name"`
+}
+
+func (k KeyslotTarget) String() string {
+	return fmt.Sprintf("(container-role: %q, name: %q)", k.ContainerRole, k.Name)
+}
+
+func (k KeyslotTarget) Validate() error {
+	if len(k.ContainerRole) == 0 {
+		return errors.New("key slot container role cannot be empty")
+	}
+	if k.ContainerRole != "system-data" && k.ContainerRole != "system-save" {
+		return fmt.Errorf(`invalid key slot container role %q, expected "system-data" or "system-save"`, k.ContainerRole)
```

Comment:
errors in `ReplaceRecoveryKey` already include the output of keyslotTarget.String(), so no need to repeat the same information

```suggestion
		return errors.New(`invalid key slot container role, expected "system-data" or "system-save"`)
```